x-pack/plugins/ingest_manager/server/services/epm/fields/field.ts

```diff
@@ -43,45 +49,133 @@ export interface Field {
+ * expandFields takes the given fields read from yaml and expands them.
  * There are dotted fields in the field.yml like `foo.bar`. These should
- * be stored as an object inside an object and is the main purpose of this
- * preprocessing.
+ * be stored as an field within a 'group' field.
  *
- * Note: This function modifies the passed field param.
+ * Note: This function modifies the passed fields array.
  */
-export function processFields(fields: Fields) {
+export function expandFields(fields: Fields) {
   fields.forEach((field, key) => {
     const fieldName = field.name;
 
-    // If the field name contains a dot, it means we need to create sub objects
+    // If the field name contains a dot, it means we need to
+    // - take the first part of the name
+    // - create a field of type 'group' with this first part
+    // - put the original field, named with the rest of the original name in the fields property of the new group field
     if (fieldName.includes('.')) {
       // Split up the name by dots to extract first and other parts
       const nameParts = fieldName.split('.');
 
       // Getting first part of the name for the new field
-      const newNameTop = nameParts[0];
-      delete nameParts[0];
+      const groupFieldName = nameParts[0];
 
       // Put back together the parts again for the new field name
-      const newName = nameParts.length === 1 ? nameParts[0] : nameParts.slice(1).join('.');
+      const restFieldName = nameParts.slice(1).join('.');
 
-      field.name = newName;
+      // keep all properties of the original field, but give it the shortened name
+      field.name = restFieldName;
 
-      // Create the new field with the old field inside
-      const newField: Field = {
-        name: newNameTop,
+      // create a new field of type group with the original field in the fields array
+      const groupField: Field = {
+        name: groupFieldName,
         type: 'group',
         fields: [field],
       };
-      // Replace the old field in the array
-      fields[key] = newField;
-      if (newField.fields) {
-        processFields(newField.fields);
+      // check child fields further down the tree
+      if (groupField.fields) {
+        expandFields(groupField.fields);
+      }
+      // Replace the original field in the array with the new one
+      fields[key] = groupField;
+    } else {
+      // even if this field doesn't have dots to expand, its child fields further down the tree might
+      if (field.fields) {
+        expandFields(field.fields);
       }
     }
   });
 }
+/**
+ * dedupFields takes the given fields and merges sibling fields with the
+ * same name together.
+ * These can result from expandFields when the input contains dotted field
+ * names that share parts of their hierarchy.
+ */
+function dedupFields(fields: Fields): Fields {
+  const dedupedFields: Fields = [];
+  fields.forEach(field => {
+    const found = dedupedFields.find(f => {
```

Comment:
This is not very efficient, for each field searching through another array of fields that potentially grows to its own length. Could you create an object that checks if a key exists instead or something like that?

Reply:
Is this an issue with the size of the input this function gets? It is at most the content of any one `fields.yml` file from a dataset in a package.

If it is an issue, can we change it in a follow-up PR? I can open an issue so we don't forget.

Reply:
It's not the input length (fields), its that its searching another list (dedupedFields) for each field to check if its duplicated.  Since this is only a single package it's not a big deal, but having another PR where we just use an object and set each field name as a property of the object and then for each field check if it's been set (instead of looping through the dupes array) would be ideal.

Reply:
Yes, but `dedupedFields` only can grow to the length of `fields`, so if I'm not mistaken this is still O(n^2), which is not terrible with the data we have in the packages.

I'll happily open another issue to fix this soon, but I don't think it should be a blocker.

Reply:
https://github.com/elastic/kibana/issues/60489